### ISSUE TITLE
Webhook validation fails

### ISSUE THREAD

**Issue by andrii2020** [sendgrid/sendgrid-python#950]
### Webhook validation fails
Hi! I'm trying to add validation to the webhook. When I'm trying to use a key, timestamp, a payload from your GitHub file with tests everything works fine. But when I generate my own key and try to use the request body which I catch via webhook.site it fails. I can provide body, key, headers which I use.  I have the last version of SendGrid installed. 
Could you please check if newly generated keys works fine for you. Thanks!

### Steps to Reproduce
1. Enable webhook in the admin panel
2. Enable validation
3. Add webhook.site link and send test request
4. Substitute received request and send it to my webhook
5. Receive fail because of bad key

**Comment by thinkingserious** [comment]
Hello @andrii2020,

One item to check is to make sure you are [using the payload in raw bytes form](https://sendgrid.com/docs/for-developers/tracking-events/getting-started-event-webhook-security-features/#verify-the-signature).

Yes, providing sample data that we can use to reproduce would be helpful. 

Thank you!

With best regards,

Elmer

**Comment by andrii2020** [comment]
x-twilio-email-event-webhook-signature  MEQCIHQkjRdbcbM/gXxai4eTHRDxsa6XyiIfXrqD6tBV7FUDAiAPZQSdHQU9O4Q/U9xA1wiDb8SppEgQbilr/Sb7sI5zCQ==x-twilio-email-event-webhook-timestamp 1604746997 | MEQCIHQkjRdbcbM/gXxai4eTHRDxsa6XyiIfXrqD6tBV7FUDAiAPZQSdHQU9O4Q/U9xA1wiDb8SppEgQbilr/Sb7sI5zCQ==
-- | --
MEQCIHQkjRdbcbM/gXxai4eTHRDxsa6XyiIfXrqD6tBV7FUDAiAPZQSdHQU9O4Q/U9xA1wiDb8SppEgQbilr/Sb7sI5zCQ==

**Comment by thinkingserious** [comment]
Another thing to check is that you are including a [trailing carriage return and newline](https://github.com/sendgrid/sendgrid-nodejs/commit/8110b173007049836455198450947780e9a70416#diff-570eee91600a9234a175a0acce452638ee2c7cc583be1a4aa75eb35ca8b9500cR19).

**Comment by andrii2020** [comment]
Yeah, everything works fine with data from your tests. but when I substitute data which I shared previously it fails.

**Comment by eshanholtz** [comment]
I'm able to reproduce the failure you're seeing with the data you shared. This issue has been added to our internal backlog to be prioritized. Pull requests and +1s on the issue summary will help it move up the backlog.

**Comment by fmaillet24** [comment]
+1
exactly the same error with exactly the same setup.

**Comment by AntovskiAntonio** [comment]
+1
I have the exact same problem. I've also tried to add the trailing carriage return and newline, as suggested by @thinkingserious, but that didn't help either. Any suggestion on how to make this work?

**Comment by nawar** [comment]
Any resolution on this?

**Comment by shwetha-manvinkurke** [comment]
Have you all tried with a trailing `\r\n` after every event? [Here](https://github.com/sendgrid/sendgrid-nodejs/blob/main/packages/eventwebhook/src/eventwebhook.spec.js#L48) is an example in node

**Comment by nawar** [comment]
@shwetha-manvinkurke yes we did in our case. The test cases in the repo work fine without issues, it seems something has changed from SendGrid side.

**Comment by corno93** [comment]
+1
I'm confused too
I've written a test that uses data from yours [here](https://github.com/sendgrid/sendgrid-python/blob/main/test/test_eventwebhook.py#L27). The test passes.

When I used my signed public verification key with integration test data triggered from sendgrid's event webhoo UI  it fails

Any idea how to resolve this?

**Comment by corno93** [comment]
I got it working by using the raw request body and decoding using latin-1. I didnt even have to use the carriage return and new line.
Here the request object is from drf 
```python
        eventwebhook = EventWebhook(settings.SENDGRID_PUBLIC_WEBHOOK_SIGNATURE)
        is_valid = eventwebhook.verify_signature(
            request.body.decode('latin-1'),
            request.headers[EventWebhookHeader.SIGNATURE],
            request.headers[EventWebhookHeader.TIMESTAMP],
        )
```

**Comment by marine-mb** [comment]
@corno93 I tried you method but it still doesn't work for me... 😔

@thinkingserious @eshanholtz Any updates on this issue ? 😊

**Comment by HendJordan** [comment]
+1

**Comment by marceloaba** [comment]
+1
Apparently multiple people are having the same issue with the signature validation. Is there any expected date to fix this issue?

**Comment by ranjanp75** [comment]
I confirm that @corno93's solution works by using **raw** request body and decoding to "latin-1" alone

**Comment by dtiesling** [comment]
Having the same issue and I've tried all of the various workarounds above to try reformatting the payload. @eshanholtz any ideas?

verification key:
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETv7beiiLEAKZNhsE9333VGkohJ2OI77+B0afOS8IASgXZESCLgnbxoaTHYvDJEWSz2S5LRKKIQfvj2h1IZVRBg==

signature:
MEUCIHAZthXvLsIgwKTW38JJPPa2czFyCa/V2ONNJIvVO9FGAiEA7pud/OufnkQQQjv9CKaiPI0cMcsMfOj0tjac6J/g9a4=

timestamp:
1715866827

payload:
[{"email": "karen.zelko@compeer.com", "event": "delivered", "ip": "168.245.98.159", "response": "250 ok:  Message 48450026 accepted", "sg_event_id": "ZGVsaXZlcmVkLTAtOTY1Njg3Mi1XVHAycnZxQlFmYVFXVGFMWlo5YnBBLTA", "sg_message_id": "WTp2rvqBQfaQWTaLZZ9bpA.recvd-7fd94c9f47-s75zg-1-66451C62-8.0", "sg_template_id": "d-4798bacd3a474f1e9461aa6261d18d10", "sg_template_name": "v1", "smtp-id": "&lt;WTp2rvqBQfaQWTaLZZ9bpA@geopod-ismtpd-6&gt;", "timestamp": 1715805283, "tls": 1}, {"email": "karen.zelko@compeer.com", "event": "processed", "send_at": 0, "sg_event_id": "cHJvY2Vzc2VkLTk2NTY4NzItV1RwMnJ2cUJRZmFRV1RhTFpaOWJwQS0w", "sg_message_id": "WTp2rvqBQfaQWTaLZZ9bpA.recvd-7fd94c9f47-s75zg-1-66451C62-8.0", "sg_template_id": "d-4798bacd3a474f1e9461aa6261d18d10", "sg_template_name": "v1", "smtp-id": "&lt;WTp2rvqBQfaQWTaLZZ9bpA@geopod-ismtpd-6&gt;", "timestamp": 1715805282}]

**Comment by dtiesling** [comment]
+1 Having the same issue and tried all of the workarounds from above. I'm attempting to handle the webhooks in a Flask view with sendgrid==6.11.0.

Does anyone have a working code snippet that formats the payload from a Flask request? Or am I missing something else in the code below?

**Latest code snippet testing the latin-1 decoding workaround**

```python
EventWebhook(public_key=public_key).verify_signature(
        payload=request.data.decode("latin-1"),
        signature=request.headers.get("X-Twilio-Email-Event-Webhook-Signature"),
        timestamp=request.headers.get("X-Twilio-Email-Event-Webhook-Timestamp"),
    )
```

**Example payload**

verification key:
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE5Qn2Y4zzFuRZGgCZagcofv4gE2mU9BtV18BO0kUVS/6lwayLXJfEoNEseO5auXrPS2ds/qpCezn1Z35qB6DBSg==

signature:
MEYCIQDuyIVG3XI1+Pe/Sq/Nr6tgSEYxOS61RCKFNHA3jsfedwIhAMFT97AX4eWDh7GJabWwVqQvZv6pcM3cypz83rbQh4tx

timestamp:
1715877013

payload:
[{"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "processed", "category": ["cat facts"], "sg_event_id": "3JMizkAyFQ2KGjExu_FXrQ==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "deferred", "category": ["cat facts"], "sg_event_id": "sF4uHYvI3Q6RYzIz94oagw==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "response": "400 try again later", "attempt": "5"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "delivered", "category": ["cat facts"], "sg_event_id": "6jZqpVNnfIu4Jio1pgDWKg==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "response": "250 OK"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "open", "category": ["cat facts"], "sg_event_id": "sTyTPHSu6yb9wT4Uguozmg==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "useragent": "Mozilla/4.0 (compatible; MSIE 6.1; Windows XP; .NET CLR 1.1.4322; .NET CLR 2.0.50727)", "ip": "255.255.255.255"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "click", "category": ["cat facts"], "sg_event_id": "IgBBSc53jusTlIZbRMHcYQ==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "useragent": "Mozilla/4.0 (compatible; MSIE 6.1; Windows XP; .NET CLR 1.1.4322; .NET CLR 2.0.50727)", "ip": "255.255.255.255", "url": "http://www.sendgrid.com/"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "bounce", "category": ["cat facts"], "sg_event_id": "UfnOu16YiTUbkTzFt2AXIA==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "reason": "500 unknown recipient", "status": "5.0.0"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "dropped", "category": ["cat facts"], "sg_event_id": "qjEbaeJnpVP-SWjXo1Sehw==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "reason": "Bounced Address", "status": "5.0.0"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "spamreport", "category": ["cat facts"], "sg_event_id": "Ww4XLyM17dl9Y5ZI9g2y4w==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "unsubscribe", "category": ["cat facts"], "sg_event_id": "lpnaz0P8OrjAu3EXZxs3Vg==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0"}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "group_unsubscribe", "category": ["cat facts"], "sg_event_id": "9aCVpaBszvmQ7gpk5LcD4w==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "useragent": "Mozilla/4.0 (compatible; MSIE 6.1; Windows XP; .NET CLR 1.1.4322; .NET CLR 2.0.50727)", "ip": "255.255.255.255", "url": "http://www.sendgrid.com/", "asm_group_id": 10}, {"email": "example@test.com", "timestamp": 1715877013, "smtp-id": "&lt;14c5d75ce93.dfd.64b469@ismtpd-555&gt;", "event": "group_resubscribe", "category": ["cat facts"], "sg_event_id": "IGR4PljOZkODcwc47CGZrg==", "sg_message_id": "14c5d75ce93.dfd.64b469.filter0001.16648.5515E0B88.0", "useragent": "Mozilla/4.0 (compatible; MSIE 6.1; Windows XP; .NET CLR 1.1.4322; .NET CLR 2.0.50727)", "ip": "255.255.255.255", "url": "http://www.sendgrid.com/", "asm_group_id": 10}]

**Comment by eric-power** [comment]
Are there any official examples of implementing this logic in python? The linked docs include a golang example, and then links to this repository which as far as I can tell do not have any examples that show setting up webhooks.

**Comment by skaaptjop** [comment]
Problem I believe might be here:
https://github.com/sendgrid/sendgrid-python/blob/3e9d4efac5d652de81fb82ae38d256aad78c00d5/sendgrid/helpers/eventwebhook/__init__.py#L47

`key.verify` wants payload in bytes, not unicode string so why not just leave the payload in bytes the whole way through? Then there is no complicated content encoding guessing along the way. Instead of having to decode the raw bytes using the correct content encoding only to encode it right back into bytes when verifying the signature... just leave it in bytes and pop the timestamp on up front.

Could look something like this:
```python
def verify_signature(self, payload: bytes, signature: str, timestamp:str, public_key=None):
        timestamped_payload_in_bytes = timestamp..encode() + payload
        decoded_signature = base64.b64decode(signature)

        key = public_key or self.public_key
        try:
            key.verify(decoded_signature, timestamped_payload_in_bytes, hashfunc=hashlib.sha256, sigdecode=sigdecode_der)
            return True
        except BadSignatureError:
            return False
```

Then you just need the raw data from the request. In flask this is just `request.get_data()`
